### ISSUE TITLE
Visualizing conv. filter through optimization

### ISSUE THREAD

**Issue by PiranjaF** [fchollet/keras#1365]
For a ConvNet it can be interesting to find the norm-bounded input that maximizes the activity of a single conv. filter as a way to [visualize the filters](http://igva2012.wikispaces.asu.edu/file/view/Erhan+2009+Visualizing+higher+layer+features+of+a+deep+network.pdf). This could be done using a black box optimization algorithm with the code from the [FAQ](http://keras.io/faq/#how-can-i-visualize-the-output-of-an-intermediate-layer).

```
# with a Sequential model
get_3rd_layer_output = theano.function([model.layers[0].input],
                                       model.layers[3].get_output(train=False))
layer_output = get_3rd_layer_output(X)
```

However, it would be a substantially easier optimization task if I had the gradient. How can I extract the gradient from the Theano expression and input it into a Python optimization library such as Scipy?

I assume that this is interesting to others as well. I'll gladly do a PR when I've got a working example.

**Comment by fchollet** [comment]
I believe some people have done that before.

Here's an example: http://nbviewer.ipython.org/github/julienr/ipynb_playground/blob/master/keras/convmnist/keras_cnn_mnist.ipynb (unfortunately outdated as we changed the Keras API since then)

A more recent example: https://github.com/AgnezIO/agnez/blob/master/examples/MNIST.ipynb

**Comment by PiranjaF** [comment]
Thanks fchollet, but those examples simply visualize the weight coefficients directly. That is not useful for anything other than the 1st layer (if even for that layer).

If I want to understand the degree to which my ConvNet is capturing the [Gabor filters](https://en.wikipedia.org/wiki/Gabor_filter) or more complex motifs at later layers such as the infamous [cat neuron](http://www.wired.com/2012/06/google-x-neural-network/), I would need to use optimization.

**Comment by PiranjaF** [comment]
Nice to see that did some work on it, @fchollet!
http://blog.keras.io/how-convolutional-neural-networks-see-the-world.html

I think you would get more interesting filter visualizations if you optimized using the norm-bounded input. [Andrew Ng's lecture nodes](web.stanford.edu/class/cs294a/sparseAutoencoder.pdf) explain that the norm-bounded input is used to get a non-trivial answer. See snippet from the notes below:

![enter image description here][1]


  [1]: http://i.stack.imgur.com/nnVdw.png

**Comment by PiranjaF** [comment]
@fchollet I'm thinking of extending the work from your blog. Do you have any insights on how it could be improved? What do you think of using the norm-bounded input?
